### PR TITLE
Added initial unit tests for the indexer

### PR DESCRIPTION
**Pull request by alex-hancock**
`test_hca_indexer.py` contains unit tests for the three bundle upload cases: correctness ([#158](https://github.com/HumanCellAtlas/data-browser/issues/158)) updating with different versions ([#159](https://github.com/HumanCellAtlas/data-browser/issues/159)), and concurrent specimen upload ([#160](https://github.com/HumanCellAtlas/data-browser/issues/160)). This pull request generally addresses data browser issue [#15](https://github.com/HumanCellAtlas/data-browser/issues/15).

`docker_es` is a brief script to retrieve the official Elasticsearch docker container and run it. I've used it in lieu of setting up my own Elasticsearch server on my machine, it might useful as we set up continuous integration for the rest of the Orange Box.

The `/testdata` folder contains all the data/metadata files referenced in `test_hca_indexer.py`, used in the mock MetadataDownloader. 

Finally, I modified imports in the dependencies of `test_hca_indexer.py` to conform to the latest repo structure.

### CODE REVIEW

File: indexer/project/hca/config.py [hunk 1]
```diff
@@ -5,9 +5,9 @@
 from elasticsearch import Elasticsearch, RequestsHttpConnection
 from requests_aws4auth import AWS4Auth
 
-from utils.base_config import BaseIndexProperties
-from utils.deployment import aws
-from utils.transformer import Transformer
+from indexer.utils.base_config import BaseIndexProperties
```
Comment (hannes-ucsc) [inline indexer/project/hca/config.py]: The `indexer` directory will go away soon. It should not be part of the package path. Set PYTHONPATH to point to `indexer` and undo these changes.

File: indexer/scripts/docker_es [hunk 2]
```diff
@@ -0,0 +1,5 @@
+#!/bin/bash
+
+docker pull docker.elastic.co/elasticsearch/elasticsearch:5.5.3
+docker run -p 9200:9200 -p 9300:9300 -e "discovery.type=single-node" -e "xpack.security.enabled=false" docker.elastic.co/elasticsearch/elasticsearch:5.5.3
```
Comment (hannes-ucsc) [inline indexer/scripts/docker_es]:
```
docker run -p 9200:9200 \
           -p 9300:9300 \
           -e "discovery.type=single-node" \
           -e "xpack.security.enabled=false" \
           docker.elastic.co/elasticsearch/elasticsearch:5.5.3
```

File: indexer/test/test_hca_indexer.py [hunk 3]
```diff
@@ -0,0 +1,271 @@
+# -*- coding: utf-8 -*-
+"""Suite for unit testing indexer.py
```
Comment (hannes-ucsc) [inline indexer/test/test_hca_indexer.py]: NL

File: indexer/test/test_hca_indexer.py [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+    metadata, data = data_pack
+    fake_event = make_fake_notification(bundle_uuid, bundle_version)
+
+    with patch.object(MetadataDownloader, 'extract_bundle') as mock_method:
+        mock_method.return_value = metadata, data
+        hca_indexer.index(fake_event)
+
+
+def make_fake_notification(uuid: str, version: str) -> Mapping[str, Any]:
+    hca_simulated_event = {
```
Comment (hannes-ucsc) [inline indexer/test/test_hca_indexer.py]: Inline variable

File: indexer/test/test_hca_indexer.py [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+                                   size=100)
+        if "project" not in entity_index:
+            for result_dict in results["hits"]["hits"]:
+                es_results.append(result_dict)
+
+    assert_func(es_results)
+    return es_results
+
+
+class Test(unittest.TestCase):
```
Comment (hannes-ucsc) [inline indexer/test/test_hca_indexer.py]: More descriptive name

File: indexer/test/test_hca_indexer.py [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+from unittest.mock import patch
+from uuid import uuid4
+
+from indexer.project.hca.indexer import Indexer
+from indexer.project.hca.config import IndexProperties
+from indexer.utils.downloader import MetadataDownloader
+from indexer.utils import eventually
+
+
+def prepare_indexer():
```
Comment (hannes-ucsc) [inline indexer/test/test_hca_indexer.py]: Any reason these can't be protected methods in the Test class?

File: indexer/test/test_hca_indexer.py [hunk 7]
```diff
@@ -0,0 +1,271 @@
...
+        spec1_pack = get_data_by_uuid(self.spec1_bundle[0])
+        spec2_pack = get_data_by_uuid(self.spec2_bundle[0])
+        specimen_list = [(self.spec1_bundle, spec1_pack),
+                         (self.spec2_bundle, spec2_pack)]
+
+        unmocked_mget = Elasticsearch.mget
+
+        def mocked_mget(self, body):
+            mget_return = unmocked_mget(self, body=body)
+            time.sleep(0.5)
```
Comment (hannes-ucsc) [inline indexer/test/test_hca_indexer.py]: Explain the intention of the sleep in a comment.

File: indexer/test/test_hca_indexer.py [hunk 8]
```diff
@@ -0,0 +1,271 @@
...
+            return mget_return
+
+        def help_index(hca_indexer, specimen_tuple):
+            mock_index(hca_indexer, specimen_tuple[0], specimen_tuple[1])
+
+        with patch.object(Elasticsearch, 'mget', new=mocked_mget):
+            with self.assertLogs(level='WARNING') as cm:
+                with ThreadPoolExecutor(max_workers=2) as executor:
+                    futures_results = executor.map(partial(help_index, self.hca_indexer), specimen_list)
```
Comment (hannes-ucsc) [inline indexer/test/test_hca_indexer.py]: The var name is misleading. tpe.map() doesn't return futures. It returns an iterable of the return values of the function.

File: indexer/test/test_hca_indexer.py [hunk 9]
```diff
@@ -0,0 +1,271 @@
...
+                    self.assertNotEqual(None, futures_results)
+                    for result in futures_results.__iter__():
```
Comment (hannes-ucsc) [inline indexer/test/test_hca_indexer.py]: Most of the time you don't want to call the magic __iter__ method, you want to use the `iter()` builtin. Then again, I'd be surprised if you need either of those.

File: indexer/test/test_hca_indexer.py [hunk 10]
```diff
@@ -0,0 +1,271 @@
...
+                        self.assertEqual(None, result)
```
Comment (hannes-ucsc) [inline indexer/test/test_hca_indexer.py]: Comparisons to None should always be done with `is` or `assertIs`

Comment (hannes-ucsc) [reply indexer/test/test_hca_indexer.py]: How about `self.assertTrue(all(r is None for r in futures_result))`

File: indexer/test/test_hca_indexer.py (hunk 8 above)
Comment (hannes-ucsc) [reply indexer/test/test_hca_indexer.py]: You can eliminate `partial` by eliminating the `hca_indexer` parameter and just using `self.hca_indexer` directly in `help_index`.

File: indexer/test/testdata/56a338fe-7554-4b5d-96a2-7df127a7640b.data [hunk 11]
```diff
@@ -0,0 +1 @@
+{'22467_6#209_1.fastq.gz': {'content-type': 'application/gzip; dcp-type=data', 'crc32c': 'c3a170ed', 'indexed': False, 'name': '22467_6#209_1.fastq.gz', 's3_etag': '7df2abd52fb511a5a0015a1d5a6d6652', 'sha1': 'd671d9c43e24a5d3efce53db9e51783628992d0a', 'sha256': 'be1413a1851354eaac53bf795f74d615b383626f4d78e7d8c7ce69806585f4c4', 'size': 1032483, 'uuid': '2d8cdf91-8162-4321-9923-e00e90fd9133', 'version': '2018-03-29T153501.427209Z'}, '22467_6#209_2.fastq.gz': {'content-type': 'application/gzip; dcp-type=data', 'crc32c': 'a39019f8', 'indexed': False, 'name': '22467_6#209_2.fastq.gz', 's3_etag': 'd478462fab81e0310d9efd4209420423', 'sha1': '94a4bfa3c6d1bf46cb424dff293311205971ad60', 'sha256': '1e0f159cf347a4a65ede1240f03101aab4885c9209a45b0d07bcd536a74b6998', 'size': 1205219, 'uuid': 'd9087c74-a41e-4702-8481-cea3ef68379c', 'version': '2018-03-29T153502.542108Z'}}
```
Comment (hannes-ucsc) [inline indexer/test/testdata/56a338fe-7554-4b5d-96a2-7df127a7640b.data]: NL missing at EOF

Can we reformat these with line breaks and indents?

File: indexer/test/test_hca_indexer.py [hunk 12]
```diff
@@ -0,0 +1,271 @@
...
+            "bundle_version": version
+        }
+    }
+    return hca_simulated_event
+
+
+def get_data_files(metadata_file, data_file):
+    data_prefix = "testdata/"
+    with open(data_prefix + metadata_file, 'r') as infile:
+        metadata = ast.literal_eval(infile.read())
```
Comment (hannes-ucsc) [inline indexer/test/test_hca_indexer.py]: What's the reason for not using `json.load()`?

**Comment by hannes-ucsc** [comment]
The branch also needs to be rebased and conflicts resolved. Don't use merge, use rebase.

**Comment by alex-hancock** [comment]
Test set-up should include initialization of Elasticsearch Docker container, test script should be removed.

Additionally, /test/testdata should be renamed /test/data.

**Comment by hannes-ucsc** [comment]
A few other things

- consider renaming `testdata` to `test/data`
- start ES in `Test.setupClass`
- mention that docker is now a development dependency

**Comment by hannes-ucsc** [comment]
Re merging/rebasing: As a reviewer I would like to see the exact series of commits that will land on the `develop` branch before I approve the PR. That's why I disabled both squashing and merging for PRs on this repository.

This PR branch does not have a meaningful history and already has merge commits in it so it would make sense that you squash it and then rebase it on develop.

**Comment by alex-hancock** [comment]
Don't merge my changes yet, I haven't updated the imports since rebasing against develop, I will update my commit.

**Comment by alex-hancock** [comment]
Imports successfully changed. Ready for review.

File: environment [hunk 13]
```diff
@@ -45,7 +45,7 @@ azul_vars=(
     AZUL_ES_VOLUME_SIZE
 
     # FIXME: not actually used outside of tests
-    
+
```
Comment (hannes-ucsc) [inline environment]: Remove this change, please.
